compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -868,15 +868,20 @@ class Namer { typer: Typer =>
     protected def addAnnotations(sym: Symbol): Unit = original match {
       case original: untpd.MemberDef =>
         lazy val annotCtx = annotContext(original, sym)
-        for (annotTree <- original.mods.annotations) {
-          val cls = typedAheadAnnotationClass(annotTree)(using annotCtx)
-          if (cls eq sym)
-            report.error(em"An annotation class cannot be annotated with iself", annotTree.srcPos)
-          else {
-            val ann = Annotation.deferred(cls)(typedAheadExpr(annotTree)(using annotCtx))
-            sym.addAnnotation(ann)
-          }
-        }
+        original.setMods: 
+          original.mods.withAnnotations :
+            for (annotTree <- original.mods.annotations) yield 
```

Comment:
In the overwhelming majority of the cases, this is going to return an equivalent list, but since it will be non-`eq`, the `with/set` methods will have to create new instances. Use `mapConverse` to avoid this issue:
```suggestion
            original.mods.annotations.mapConserve { annotTree =>
```
(or whatever is its correct syntax)